### PR TITLE
Added `-[IGListAdapter visibleCellsForObject:]` API

### LINKED ISSUE #437: Add -[IGListAdapter visibleCellsForObject:] API
Got a request internally for this API. It's doable now, but more work than necessary:

- Get the section of an object with `sectionForObject:`
- Create an `NSIndexPath`
- Get the cell with `cellForRowAtIndexPath:`

Would be nice to just grab all visible cells that exist for a given object.

### PR DESCRIPTION
**Pull request by Sherlouk**
## Changes in this pull request

- Added `-[IGListAdapter visibleCellsForObject:]` API

## Approach

1. Get all visible cells
2. Get the section for the object
3. Filter cells where indexPath.section of cell is equal to the object
4. Return

## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added test(s), an experiment, or detailed why my change isn't tested.
- [ ] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [x] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

-------------

Purely additive, can go in 2.2.0 release? Hold off on changelog until agreed.
Wanting to learn some more ObjC so gave this a shot, would appreciate some nit picking/feedback

Closes #437

### CODE REVIEW

**Comment by Sherlouk** [comment]
Turns out there's already a method for `-[IGListCollectionContext visibleCellsForSectionController:]` 

https://instagram.github.io/IGListKit/Protocols/IGListCollectionContext.html#/c:objc(pl)IGListCollectionContext(im)visibleCellsForSectionController:

https://github.com/Instagram/IGListKit/blob/master/Source/IGListAdapter.m#L777-L788

Considering you can quite easily get the section controller from object using `-[IGListAdapter sectionControllerForObject:]` it doesn't seem too difficult to do this already?

@rnystrom Suggestions?

**Comment by rnystrom** [comment]
@Sherlouk doh. Ya, maybe we should just recommend people use that API instead...

**Comment by Sherlouk** [comment]
@rnystrom Was your API request, if you think the need is no longer necessary then close this and that issue 😄

**Comment by rnystrom** [comment]
Lol I know, forgot that API existed. Might still be valuable to cut out the middle-layer steps. Could also return early if `sectionController` is `nil`.

**Comment by Sherlouk** [comment]
Sorry, lost track here...

We saying in the old API to return early?
What steps do we want to take here

(Also pardon the test name 😂, only just realised I didn't change it)

**Comment by rnystrom** [comment]
The old API will end up [asserting](https://github.com/Instagram/IGListKit/blob/master/Source/IGListAdapter.m#L340) if the section controller is nil. If you used this method by first getting the object, then the SC (in ObjC) you might assert:

```objc
id controller = [self.adapter sectionControllerForObject:object]; // can be nil
NSArray *cells = [self.adapter visibleCellsForSectionController:controller]; // assert if nil
```

Which means all your callsites have to look like:

```objc
id controller = [self.adapter sectionControllerForObject:object]; // can be nil
if (controller != nil) {
  NSArray *cells = [self.adapter visibleCellsForSectionController:controller]; // assert if nil
}
```

Which is kind of gross. Swift will be just as bad. Options are:

- Remove the assert
- Add `visibleCellsForObject:` method that returns early if a section controller isn't found for the object

We should also add a unit test that sends `nil` to `visibleCellsForSectionController:` even tho the param marked `nonnull`, in ObJC-land that can still happen and the compiler wont help you.

**Comment by rnystrom** [comment]
@Sherlouk ok thought about it more. Let's do option 2 and move forward w/ this for 2.2 if you think that's doable 😄

File: Source/IGListAdapter.h [hunk 1]
```diff
@@ -195,6 +195,15 @@ IGLK_SUBCLASSING_RESTRICTED
 - (NSArray *)visibleObjects;
 
 /**
+ An unordered array of the currently visible cells for a given object.
+ 
+ @param item An object in the list
+ 
+ @return An array of collection view cells.
+ */
+- (NSArray<UICollectionViewCell *> *)visibleCellsForObject:(id)item;
```
Comment (rnystrom) [inline Source/IGListAdapter.h]: nit: `object` as param name

File: Source/IGListAdapter.m [hunk 2]
```diff
@@ -414,6 +414,23 @@ - (NSArray *)visibleObjects {
...
+- (NSArray<UICollectionViewCell *> *)visibleCellsForObject:(id)item {
+    IGAssertMainThread();
+    IGParameterAssert(item != nil);
+    
+    NSArray<UICollectionViewCell *> *visibleCells = [self.collectionView visibleCells];
+    NSInteger section = [self.sectionMap sectionForObject:item];
+    IGAssert(section != NSNotFound, @"Section not found for object %@", item);
+    
+    NSPredicate *controllerPredicate = [NSPredicate predicateWithBlock:^BOOL(UICollectionViewCell* cell, NSDictionary* bindings) {
+        NSIndexPath *indexPath = [self.collectionView indexPathForCell:cell];
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: nit: calling `self.collectionView` inside this block will execute an extra `objc_msgSend(...)` every execution. Instead let's capture the collection view in a local variable that the block uses:
```objc
UICollectionView *collectionView = self.collectionView;
NSPredicate *predicate = [... ^{
  NSIndexPath *path = [collectionView indexPathForCell:cell];
}];
```

File: Source/IGListAdapter.m [hunk 3]
```diff
@@ -414,6 +414,23 @@ - (NSArray *)visibleObjects {
     return [visibleObjects allObjects];
 }
 
+- (NSArray<UICollectionViewCell *> *)visibleCellsForObject:(id)item {
+    IGAssertMainThread();
+    IGParameterAssert(item != nil);
+    
+    NSArray<UICollectionViewCell *> *visibleCells = [self.collectionView visibleCells];
+    NSInteger section = [self.sectionMap sectionForObject:item];
+    IGAssert(section != NSNotFound, @"Section not found for object %@", item);
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: Actually let's remove this. A use might be holding a ref to an expired object. I think it's ok to return an empty array in that case.

File: Source/IGListAdapter.m [hunk 4]
```diff
@@ -414,6 +414,23 @@ - (NSArray *)visibleObjects {
     return [visibleObjects allObjects];
 }
 
+- (NSArray<UICollectionViewCell *> *)visibleCellsForObject:(id)item {
+    IGAssertMainThread();
+    IGParameterAssert(item != nil);
+    
+    NSArray<UICollectionViewCell *> *visibleCells = [self.collectionView visibleCells];
+    NSInteger section = [self.sectionMap sectionForObject:item];
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: nit: `const`

**Comment by Sherlouk** [comment]
Option 2 is good, want to raise a new issue to track progress on amending old functionality?
Will add another unit test as well testing if an object is not found that it returns empty

File: Source/IGListAdapter.m [hunk 5]
```diff
@@ -414,6 +414,23 @@ - (NSArray *)visibleObjects {
...
+    IGAssert(section != NSNotFound, @"Section not found for object %@", item);
+    
+    NSPredicate *controllerPredicate = [NSPredicate predicateWithBlock:^BOOL(UICollectionViewCell* cell, NSDictionary* bindings) {
+        NSIndexPath *indexPath = [self.collectionView indexPathForCell:cell];
+        return indexPath.section == section;
+    }];
+    
+    NSArray<UICollectionViewCell *> *filteredCells = [visibleCells filteredArrayUsingPredicate:controllerPredicate];
+    return filteredCells;
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: nit: IMO just `return [visibleCells filteredArrayUsingPredicate:controllerPredicate];`

**Comment by Sherlouk** [comment]
All valid comments, will work on it tonight!

**Comment by Sherlouk** [comment]
@rnystrom Just updated, should address all of your points!

Not sure if this is the best code to return an empty array, but it does seem to work?
```objc
return [NSArray new];
```
